### PR TITLE
Sebastienlevert/fix chaos handler random

### LINKED ISSUE #427: ChaosHandler cannot be used in interactive scenarios
# Bug Report

## Prerequisites

-   [x] Can you reproduce the problem?
-   [x] Are you running the latest version?
-   [x] Are you reporting to the correct repository?
-   [x] Did you perform a cursory search?

## Description

When using ChaosHandler, it should work in interactive scenarios (not only Unit Tests). Right now, if it fails once, it will always fail, which creates a wrong experience in testing throttling handling.

## Steps to Reproduce

1. Inject a ChaosHandler in a list of middleware
2. If the first requests gets throttled, then it will always fail. Making testing an interactive scenario impossible

**Expected behavior:** [What you expected to happen]

Every new request should get randomized and potentially succeed.

**Actual behavior:** [What actually happened]

It's always failing if the initial random identifies the request as a 429.

## Additional Context

Add any other context about the problem here..

## Usage Information

SDK Version - [SDK version you are using]

-   [x] Node (Check, if using Node version of SDK)

> Node Version - 10

-   [x] Browser (Check, if using Browser version of SDK)

> Browser Name - Edge

> Version - 89.0
 
[AB#9063](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/9063)

### PR DESCRIPTION
**Pull request by sebastienlevert**
## Motivation

Brings the ChaosHandler to support interactive scenarios. Will allow for randomized requests to fail but also retries attemps to both fail and succeed based on the percentage provided.

## Closing issues

Fixes #427 

## Types of changes

-   [x] Bug fix (non-breaking change which fixes an issue)
-   [] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

-   [x] I have read the **CONTRIBUTING** document.
-   [x] My code follows the code style of this project.
-   [ ] My change requires a change to the documentation.
-   [ ] I have updated the documentation accordingly.
-   [x] I have added tests to cover my changes.
-   [x] All new and existing tests passed.

### CODE REVIEW

File: src/middleware/ChaosHandler.ts [hunk 1]
```diff
@@ -141,10 +141,12 @@ export class ChaosHandler implements Middleware {
 	 */
 	private async sendRequest(chaosHandlerOptions: ChaosHandlerOptions, context: Context): Promise<void> {
 		this.setStatusCode(chaosHandlerOptions, context.request as string, context.options.method as RequestMethod);
-		if (!chaosHandlerOptions.statusCode) {
-			await this.nextMiddleware.execute(context);
-		} else {
+		if (Math.floor(Math.random() * 100) < chaosHandlerOptions.chaosPercentage) {
```
Comment (nikithauc) [inline src/middleware/ChaosHandler.ts:144]: This condition will always be checked even when the user selects 'ChaosStrategy.MANUAL'.  and  as ChaosPercentage defaults to 10 -> 
`this.chaosPercentage = chaosPercentage !== undefined ? chaosPercentage : 10;`

ChaosPercentage indicates the percentage of randomness/chaos in the handle. 

Setting the ChaosPercentage when using the manual strategy might be a confusing convention. 

I feel we should check for the chaos strategy as well and determine the condition.

**Comment by nikithauc** [comment]
@sebastienlevert  Thank you for your contribution!

File: test/common/middleware/ChaosHandler.ts [hunk 2]
```diff
@@ -276,5 +276,17 @@ describe("ChaosHandler.ts", () => {
 			};
 			assert.isDefined(tempChaosHandlerManual["execute"](cxt));
 		});
+
+		it("Should return response for Manual Request Level case 100%", async () => {
```
Comment (baywet) [inline test/common/middleware/ChaosHandler.ts:280]: no await in the body, remove the async

**Review by baywet — CHANGES_REQUESTED** [review]
thanks for the contribution @sebastienlevert ! a few minor suggestions

File: test/common/middleware/ChaosHandler.ts (hunk 2 above)
Comment (sebastienlevert) [reply test/common/middleware/ChaosHandler.ts:280]: They are all using async without await... Want me to clear all of them in this file?

File: src/middleware/ChaosHandler.ts (hunk 1 above)
Comment (sebastienlevert) [reply src/middleware/ChaosHandler.ts:144]: Makes sense. I'll do both!

File: test/common/middleware/ChaosHandler.ts (hunk 2 above)
Comment (baywet) [reply test/common/middleware/ChaosHandler.ts:280]: sure. thanks to taking this on.

**Review by baywet — APPROVED** [review]

File: src/middleware/ChaosHandler.ts [hunk 3]
```diff
@@ -141,10 +141,10 @@ export class ChaosHandler implements Middleware {
 	 */
 	private async sendRequest(chaosHandlerOptions: ChaosHandlerOptions, context: Context): Promise<void> {
 		this.setStatusCode(chaosHandlerOptions, context.request as string, context.options.method as RequestMethod);
-		if (!chaosHandlerOptions.statusCode) {
-			await this.nextMiddleware.execute(context);
-		} else {
+		if (chaosHandlerOptions.chaosStrategy === ChaosStrategy.MANUAL || Math.floor(Math.random() * 100) < chaosHandlerOptions.chaosPercentage) {
```
Comment (nikithauc) [inline src/middleware/ChaosHandler.ts:144]:
```suggestion
		if ((chaosHandlerOptions.chaosStrategy === ChaosStrategy.MANUAL && !this.nextMiddleware) || Math.floor(Math.random() * 100) < chaosHandlerOptions.chaosPercentage) {
```

**Pull request merged**